### ISSUE TITLE
Replace various `type M env m`s with common intersection

### ISSUE THREAD

**Issue by sjakobi** [commercialhaskell/stack#1744]
The following files contain definitions of some `type M env m` which largely seem to be copies of each other:

```
src/Stack/Build.hs
src/Stack/Build/Execute.hs
src/Stack/Build/Installed.hs
src/Stack/Docker.hs
src/Stack/Nix.hs
src/Stack/SDist.hs
```

This leads to some redundancy and to less informative haddocks ([example](http://haddock.stackage.org/lts-5.1/stack-1.0.2/Stack-Build.html)) because none of these types are exported.

I was wondering whether there would be a point in creating an alias of the intersection of these type constraints and exporting it (from `Stack.Types.StackT` maybe), so it shows in the haddocks. One idea for a name would be `Common`.

This idea conflicts with #1297.

**Comment by mgsloan** [comment]
Makes sense.  I think the reasoning for having module local definitions of it is to make it easy to add a constraint to everything in the module.  I also find it hard to follow, though.

I find the concrete approach to be the easiest to read.  It provides better type errors, and better performance (nothing that really matters, though).  It does sacrifice some flexibility, but I think in most cases we aren't using that flexibility.  Then again, perhaps this is an aspect of stack's design that we don't want to sacrifice.  Thoughts, @snoyberg @chrisdone ?

**Comment by sjakobi** [comment]
I'm rather leaning towards the approach proposed in #1297 now.
